packages/tanstack-react-query/src/internals/createOptionsProxy.ts

```diff
@@ -0,0 +1,309 @@
+  > &
+    DecorateQueryKeyable;
+
+export interface TRPCOptionsProxyOptionsBase {
+  queryClient: QueryClient | (() => QueryClient);
+  overrides?: {
+    mutations?: MutationOptionsOverride;
+  };
+}
+
+export interface TRPCOptionsProxyOptionsInternal<TRouter extends AnyRouter> {
+  router: TRouter;
+  ctx:
+    | inferRouterContext<TRouter>
+    | (() => MaybePromise<inferRouterContext<TRouter>>);
+}
+
+export interface TRPCOptionsProxyOptionsExternal<TRouter extends AnyRouter> {
+  client: TRPCUntypedClient<TRouter> | CreateTRPCClient<TRouter>;
+}
+
+export type TRPCOptionsProxyOptions<TRouter extends AnyRouter> =
+  TRPCOptionsProxyOptionsBase &
+    (
+      | TRPCOptionsProxyOptionsInternal<TRouter>
+      | TRPCOptionsProxyOptionsExternal<TRouter>
+    );
+
+type UtilsMethods =
+  | keyof DecorateQueryProcedure<any>
+  | keyof DecorateMutationProcedure<any>
+  | keyof DecorateSubscriptionProcedure<any>;
+
+function getQueryType(method: UtilsMethods) {
+  const map: Partial<Record<UtilsMethods, QueryType>> = {
+    queryOptions: 'query',
+    infiniteQueryOptions: 'infinite',
+    subscriptionOptions: 'any',
+    mutationOptions: 'any',
+  };
+
+  return map[method];
+}
+
+export function createTRPCOptionsProxy<TRouter extends AnyRouter>(
+  opts: TRPCOptionsProxyOptions<TRouter>,
+): TRPCOptionsProxy<TRouter> {
+  const callIt = (type: ProcedureType): any => {
+    return (path: string, input: unknown, trpcOpts: TRPCRequestOptions) => {
+      if ('router' in opts) {
+        return Promise.resolve(unwrapLazyArg(opts.ctx)).then((ctx) =>
+          callProcedure({
+            procedures: opts.router._def.procedures,
+            path: path,
+            getRawInput: async () => input,
+            ctx: ctx,
+            type: type,
+            signal: undefined,
+          }),
+        );
+      }
+
+      const untypedClient =
+        opts.client instanceof TRPCUntypedClient
+          ? opts.client
+          : getUntypedClient(opts.client);
+
+      return untypedClient[type](path, input, trpcOpts);
+    };
+  };
+
+  return createRecursiveProxy(({ args, path: _path }) => {
+    const path = [..._path];
+    const utilName = path.pop() as UtilsMethods;
+    const [arg1, arg2] = args as any[];
+
+    const queryType = getQueryType(utilName);
+    const queryKey = getQueryKeyInternal(path, arg1, queryType ?? 'any');
+
+    const contextMap: Record<UtilsMethods, () => unknown> = {
+      '~types': undefined as any,
+
+      mutationKey: () => {
+        return getMutationKeyInternal(path);
+      },
+      queryKey: () => {
+        return queryKey;
+      },
+      queryFilter: (): QueryFilters => {
+        return {
+          queryKey: queryKey,
+        };
+      },
+      infiniteQueryOptions: () => {
+        return trpcInfiniteQueryOptions({
+          opts: arg2,
+          path,
+          queryClient: opts.queryClient,
+          queryKey: queryKey,
+          query: callIt('query'),
+        });
+      },
```

Comment:
I wonder if we should make all of these independent factory functions. A bit more boilerplate but then tree-shaking etc works out of the box so if you don't use e.g. subscriptions none of that code is imported

Reply:
What!? You want 1 proxy per function? 

I don't think there are many bytes here to care about treeshaking

Reply:
perhaps, if we move the function to be standalone:

```ts
import { useQuery } from "@tanstack/react-query";
import { queryOptions } from "@trpc/tanstack-react-query";

const proxy = createProxy<AppRouter>(..)

function Comp() {
  const query = useQuery(
    queryOptions(
      proxy.post.list(input, { ...opts })
    )
  )
}
```

but idk, hurts discoverability for what I imagine isn't much gains in total bundle size